src/psyclone/dynamo0p3.py

```diff
@@ -1520,10 +1554,26 @@ def _declare_unique_extent_vars(self, parent):
         api_config = Config.get().api_conf("dynamo0.3")
 
         if self._unique_extent_vars:
-            parent.add(DeclGen(parent, datatype="integer",
-                               kind=api_config.default_kind["integer"],
-                               entity_decls=self._unique_extent_vars,
-                               intent="in"))
+            for arg in self._kern_args:
```

Comment:
As discussed, I suggest restructuring this as
```
if self._kernel:
    for arg in self._kern_args:
        if arg.descriptor.stencil['type'] == "cross2d":
            ... code for CROSS2D
        else:
            ... code for other stencils
elif self._invoke:
    parent.add(DeclGen(parent, datatype="integer",
        kind=api_config.default_kind["integer"],
        ....
else:
    raise InternalError(....
```
See e.g. `def _unique_extent_vars` for an example (please also add the relevant test for `InternalError`).

Reply:
Adding an internal error here appears to be redundant (as i found while writing the test for it) so I've removed it, the call to `self._unique_extent_vars` on line 1555 calls a method which has the same internal error check (line 1538) so the internal error is raised before the code gets to the `if self._kern_args` on line 1556.